geoopt/manifolds/symmetric_positive_definite.py

```diff
@@ -0,0 +1,385 @@
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1}
+        """
+        return self._funcm(x, torch.reciprocal)
+
+    def _inv_sqrtm1(self, x: torch.Tensor) -> torch.Tensor:
+        """Symmetric matrix inverse square root.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1/2}
+        """
+        return self._funcm(x, lambda tensor: torch.reciprocal(torch.sqrt(tensor)))
+
+    def _inv_sqrtm2(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Symmetric matrix inverse square root, with square root return also.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            x^{-1/2}, sqrt(x)
+        """
+        e, v = torch.symeig(x, eigenvectors=True)
+        sqrt_e = torch.sqrt(e)
+        inv_sqrt_e = 1 / sqrt_e
+        return (
+            v @ torch.diag_embed(inv_sqrt_e) @ v.transpose(-1, -2),
+            v @ torch.diag_embed(sqrt_e) @ v.transpose(-1, -2),
+        )
+
+    def _sym(self, x: torch.Tensor) -> torch.Tensor:
+        r"""Make matrix symmetric.
+
+        .. math::
+
+            \frac{A + A^T}{2}
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            matrix to symmetrize
+
+        Returns
+        -------
+        torch.Tensor
+            symmetric matrix
+        """
+        return (x + x.transpose(-1, -2)) / 2
+
+    _dist_doc = """
+        Parameters
+        ----------
+        x : torch.Tensor
+            point on the manifold
+        y : torch.Tensor
+            point on the manifold
+        keepdim : bool
+            keep the last dim?
+
+        Returns
+        -------
+        torch.Tensor
+            distance between two points
+        """
+
+    def _affine_invariant_metric(
+        self, x: torch.Tensor, y: torch.Tensor, keepdim=False
+    ) -> torch.Tensor:
+        r"""Affine Invariant Metric distance.
+
+        {}
+
+        References
+        ----------
+        A Riemannian framework for tensor computing. 2006.
+        """.format(
+            self._dist_doc
+        )
+        inv_sqrt_x = self._inv_sqrtm1(x)
+        return 0.5 * torch.norm(
+            self._logm(inv_sqrt_x @ y @ inv_sqrt_x), dim=[-1, -2], keepdim=keepdim
+        )
+
+    def _stein_metric(
+        self, x: torch.Tensor, y: torch.Tensor, keepdim=False
+    ) -> torch.Tensor:
+        r"""Stein Metric distance.
+
+        {}
+
+        References
+        ----------
+        A new metric on the manifold of kernel matrices with application to matrix geometric means. 2012.
+        """.format(
+            self._dist_doc
+        )
+
+        def log_det(tensor: torch.Tensor) -> torch.Tensor:
+            return torch.log(torch.det(tensor))
+
+        if keepdim:
```

Comment:
keep_dim is an important parameter, let's just expand collapsed dims in this case?